### PR TITLE
[HW04] Implementation of the interaction between students and mentors

### PR DESCRIPTION
**Pull request by miseinen**
- _**Member class**_ represents a course participant. 
- _**Student class**_ represents a member who may solve task (derived from Member).
- _**Mentor**_ is a member who may add/review tasks (derived from Member).
- _**Homework class**_ represents  tasks.
-  _**Answer class**_ contains student and his answer for homework. 
- _**Homeworks Manager**_ manages homework state and appropriate process when state of homework changes. It includes a _**Singleton**_ because Homeworks Manager must be only one. Also it includes a _**Members Manager**_ allowing to manage members access to do/review homeworks and containing lists of reviewers and solvers. 
- _**Homework Participant**_ represent member with certain homework.
- _**Homework State**_ module represent the homework status and provide appropriate homework process. When homework status changes all reviewers and solvers get notifications.  Homework may have the following states: _new, done, rejected, accepted_.
- _**Work Demonstration**_ created for test and demonstration of app possibility.
### Demonstration example:
part#1
part#2
part3

### CODE REVIEW

**Comment by vladhilko** [comment]
48 commits and 16 files  :scream: Did you decide to train all patterns together? :upside_down_face:  Why don't you like KISS principle? 😅 I need a description from your side 

Screenshot 2021-05-06 at 21 55 29

**Comment by miseinen** [comment]
> 48 commits and 16 files scream Did you decide to train all patterns together? upside_down_face Why don't you like KISS principle? sweat_smile I need a description from your side
> 
> Screenshot 2021-05-06 at 21 55 29

Done. But I don't know if it's written by right way.

**Comment by miseinen** [comment]
> I need a description from your side
>
I also added screenshots demonstrating work process.

File: HW04/ekaterina_balabanovich/lib/work_demonstration.rb [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+require_relative 'student'
+require_relative 'mentor'
+require_relative 'homework'
+require_relative 'notification'
+
+class WorkDemonstration
+  def run
+    mentor_one = create_mentor('Ivan', 'Stepanov')
+    mentor_two = create_mentor('Trevor', 'Dutch')
```
Comment (vladhilko) [inline HW04/ekaterina_balabanovich/lib/work_demonstration.rb:11]: Minor. We usually use numbers for such variables
```suggestion
    mentor1 = create_mentor('Ivan', 'Stepanov')
    mentor2 = create_mentor('Trevor', 'Dutch')
```

File: HW04/ekaterina_balabanovich/lib/work_state/work_state_rejected.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class WorkStateRejected < WorkState
+  def process(members, homework)
+    WorksManager.instance.notify(members: members, task: homework, status: :reject)
+  end
+end
```
Comment (vladhilko) [inline HW04/ekaterina_balabanovich/lib/work_state/work_state_rejected.rb:7]: We usually do something like this if we want to represent such structure 
```suggestion
module WorkStates
  class Rejected < Base
    def process(members, homework)
      WorksManager.instance.notify(members: members, task: homework, status: :reject)
    end
  end
end

WorkState::Rejected.new
```

File: HW04/ekaterina_balabanovich/lib/work_state/work_state.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+class WorkState
+  def process(members, homework); end
+end
```
Comment (vladhilko) [inline HW04/ekaterina_balabanovich/lib/work_state/work_state.rb:5]: I would prefer to call it `Base` or `Abstract` to represent a real value of this class
```suggestion
module WorkState
  class Base
    def process(members, homework); end
  end
end
```

File: HW04/ekaterina_balabanovich/lib/task_solver.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+module TaskSolver
```
Comment (vladhilko) [inline HW04/ekaterina_balabanovich/lib/task_solver.rb:3]: I don't see any sense to create such a module if you use it only once 🤷‍♂️ It just increases complexity

File: HW04/ekaterina_balabanovich/lib/task_solver.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+  attr_reader :homeworks
+
+  def get_homework(homework)
+    @homeworks << homework
+  end
+
+  def delete_homework(homework)
+    @homeworks.delete(homework)
+  end
```
Comment (vladhilko) [inline HW04/ekaterina_balabanovich/lib/task_solver.rb:12]: If it's possible would be better not to use instance variables
```suggestion
  def get_homework(homework)
    homeworks << homework
  end

  def delete_homework(homework)
    homeworks.delete(homework)
  end
```

File: HW04/ekaterina_balabanovich/lib/task_solver.rb [hunk 6]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+module TaskSolver
+  attr_reader :homeworks
+
+  def get_homework(homework)
```
Comment (vladhilko) [inline HW04/ekaterina_balabanovich/lib/task_solver.rb:6]: Why do you call it `get_homework` if you `add_homework` inside? 🤔

File: HW04/ekaterina_balabanovich/lib/notification_receiver.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+module NotificationReceiver
```
Comment (vladhilko) [inline HW04/ekaterina_balabanovich/lib/notification_receiver.rb:3]: The same comments as for `TaskSolver`

File: HW04/ekaterina_balabanovich/lib/work_demonstration.rb [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+  def show_homeworks(member)
+    puts_with_delimiter("#{member.class} #{member.surname}\'s homeworks (to work/to check):")
+    member.homeworks.each { |a| puts a.title }
+  end
+
+  def show_answers(member)
+    puts_with_delimiter("#{member.class} #{member.surname} see answers")
+    member.homeworks.each do |a|
+      puts a.title
+      a.answers.each { |k, v| puts "#{k.surname}: #{v}" }
```
Comment (vladhilko) [inline HW04/ekaterina_balabanovich/lib/work_demonstration.rb:96]: Can we use the whole name for the variables? 🤔 It's not clear what is going on here 
```suggestion
    member.homeworks.each do |homework|
      puts homework.title
      homework.answers.each { |student, answer| puts "#{student.surname}: #{answer}" }
```

Comment (vladhilko) [reply HW04/ekaterina_balabanovich/lib/work_demonstration.rb:96]: This comment related to the whole task we need to fix it everywhere 😉

File: HW04/ekaterina_balabanovich/lib/homework.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+  def initialize(title, description)
+    @title = title
+    @description = description
+    @answers = Hash.new
+  end
+
+  def solve(answer, student)
+    answers[student] = answer
+  end
```
Comment (vladhilko) [inline HW04/ekaterina_balabanovich/lib/homework.rb:15]: This is not a good practice to use Hash for this kind of things https://refactoring.guru/ru/smells/primitive-obsession would be better to create a separated class or just a new `Struct` https://ruby-doc.org/core-2.5.0/Struct.html
```suggestion
    @answers = []
  end
  
  Answer = Struct.new(:student, :answer)

  def solve(answer, student)
    answers << Answer.new(student, answer)
  end

```

File: HW04/ekaterina_balabanovich/lib/work_demonstration.rb (hunk 8 above)
Comment (vladhilko) [inline HW04/ekaterina_balabanovich/lib/work_demonstration.rb:96]: If you create a new class/struct `Answer` than you could just add something like this 
```suggestion
      homework.answers.each { |answer| answer.show }
      
      # answer
          def show
            puts "#{student.surname}: #{answer}"
          end
```

File: HW04/ekaterina_balabanovich/lib/notification.rb [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+class Notification
+  attr_reader :homework, :note
+
+  def initialize(homework)
+    @homework = homework
+    @note = { new: "You got new homework: '#{homework.title}' to work",
+              done: "You got '#{homework.title} to check",
+              reject: "Your '#{homework.title}' have been rejected",
+              accept: "Your '#{homework.title} have been accepted" }
+  end
```
Comment (vladhilko) [inline HW04/ekaterina_balabanovich/lib/notification.rb:12]: I would split them a little bit 
```suggestion
    @note = { new: new_message,
              done: done_message,
              reject: reject_message,
              accept: accept_message }
  end

  def new_message
    "You got new homework: '#{homework.title}' to work"
  end

  def done_message
    "You got '#{homework.title} to check"
  end

  def reject_message
    "Your '#{homework.title}' have been rejected"
  end

  def accept_message
    "Your '#{homework.title} have been accepted"
  end
```

File: HW04/ekaterina_balabanovich/lib/works_manager.rb [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+class WorksManager
+  include MembersManager
+  include Singleton
+
+  attr_reader :homeworks, :state, :notifications
+
+  def initialize
+    @reviewers = Hash.new
+    @solvers = Hash.new
+    @notifications = Hash.new
```
Comment (vladhilko) [inline HW04/ekaterina_balabanovich/lib/works_manager.rb:20]: The same problem here https://refactoring.guru/ru/smells/primitive-obsession if you know how to rewrite it using `class/struct` or just array that would be nice 🙂

**Comment by vladhilko** [comment]
Second round 🙂

File: HW04/ekaterina_balabanovich/lib/member.rb [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+  def initialize(name, surname)
+    @name = name
+    @surname = surname
+    @homeworks = []
+    @notifications = []
+  end
+
+  def get_notifications(notification)
+    notifications << notification
+  end
```
[the remark this answers is not in the document]
Comment (miseinen) [reply HW04/ekaterina_balabanovich/lib/member.rb:15]: I fixed get_homework)

**Comment by miseinen** [comment]
> Second round slightly_smiling_face

It was extremely useful round :D

File: HW04/ekaterina_balabanovich/lib/work_state/new.rb [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+module WorkState
+  class New < Base
+    def process(members, homework)
+      WorksManager.instance.notify(members: members, task: homework, status: :new)
+      if members.is_a? Array
+        members.each { |member| member.add_homework(homework) }
+      else
+        members.add_homework(homework)
+      end
```
Comment (vladhilko) [inline HW04/ekaterina_balabanovich/lib/work_state/new.rb:11]: This is a bad approach. If you call something `members` then we should always receive an array, but I don't know how difficult will be to wrap all these single cases into an array 
```suggestion
      members.each { |member| member.add_homework(homework) }
```

**Review by vladhilko — COMMENTED** [review]
3rd round 🙂

**Review by vladhilko — APPROVED** [review]
Nice job 👍

**Pull request merged**